power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -59,20 +60,64 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
 
 template <symmetry_tag sym>
 inline void calculate_source_result(IdxRange const& sources, Idx bus_number, YBus<sym> const& y_bus,
-                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output) {
-    for (Idx const source : sources) {
-        ComplexValue<sym> const u_ref{input.source[source]};
-        ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
-        output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    ComplexValue<sym> const& i_load_gen_bus) {
+    ComplexValue<sym> const i_inj_t = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    std::vector<Idx> sources_acc(sources.size());
+    std::ranges::transform(sources, sources_acc.begin(), [&](Idx const source) { return source; });
+    if (sources_acc.empty()) {
+        ; // Do nothing
+    } else if (sources_acc.size() == 1) {
+        output.source[sources_acc[0]].i = i_inj_t;
+        output.source[sources_acc[0]].s = output.u[bus_number] * conj(output.source[sources_acc[0]].i);
+    } else {
```

Comment:
```suggestion
    if (sources_acc.size() == 1) {
        output.source[sources_acc[0]].i = i_inj_t;
        output.source[sources_acc[0]].s = output.u[bus_number] * conj(output.source[sources_acc[0]].i);
    } else if (!sources_acc.empty()) {
```

Reply:
I was suggested to make it explicit that nothing must be done when there are no sources, but I feel that your way accomplishes this more elegantly. Thanks!